[GSoC] Implement Compiler of a Simple Language

evansb opened the pull request:
Main Features
- [x] Binary Expression
- [x] Setup Typing
- [x] Let binding
- [x] If Expression
- [x] Lamdba and Apply
- [x] External Math function
- [x]  Capture in lambda

Extra Features
- [x] Eta expansion for foreign functions
- [x] Recursive function

Apologise for the delay,
I've been busy with school project, but I'll be free from now until 1 April.

evansb committed:
Ignore IDE files

evansb committed:
Move test helpers to separate file to declutter the test files

evansb committed:
Initial Compiler Implementation

evansb committed:
Refactor main compiler class by splitting into three files

evansb committed:
Finish Binary Operation with Test

sjrd commented on .gitignore:
```diff
@@ -1 +1,3 @@
 target/
+
+.idea/
```
IDE-specific ignores should not be put in a project's `.gitignore` file. They should be in [your global gitignore](https://help.github.com/articles/ignoring-files/#create-a-global-gitignore).

evansb replied:
Noted, thanks

sjrd commented on src/main/scala/calc/NodeCompiler.scala:
```diff
@@ -0,0 +1,31 @@
+import org.scalajs.core.ir.{Trees => irt}
+
+object NodeCompiler {
+  def apply(f: Tree => irt.Tree): NodeCompiler = {
```
That is the first time I ever see anything like that. It's going to be crazy inefficient on any non-trivial programs.

At least, instead of catching `CompileError`s, you should use a `PartialFunction` and chain things with `orElse`. But even that is a very unorthodox way of writing a compiling. Why did you not use a simple pattern matching? More efficient and more concise, what is there not to love?

evansb replied:
Yes, I agree that a simple pattern matching will be more efficient and sufficient for this simple language, what I wanted to show you here is my attempt to consider about the modularity of the compiler, i.e when the rules become large, a single pattern matching will be unmaintainable, I think.
I saw this approach of breaking compilers into smaller "node compilers" in CoffeeScript compiler.

I'll change it to use pattern matching, thanks for the feedback

evansb committed:
Move IDE .gitignore to user's global

evansb committed:
Setup Typer phase and simplify design

evansb commented:
Changed to simpler pattern matching.

sjrd commented on src/main/scala/calc/Tree.scala:
```diff
@@ -11,16 +16,42 @@ final case class Literal(value: Double)(implicit val pos: Position) extends Tree
+
+sealed abstract class TreeT {
+  def pos: Position
+  def getType(): Type
```
That's a bit Java-ish ;) In Scala we don't use `getX`-style getters. We just use `def tpe: Type` (without `get` nor `()`) or possibly `val tpe: Type`.

evansb replied:
Yeah, I was unsure what to use since `type` is a reserved keyword.

I'll use `tpe` as you suggested.

sjrd commented on src/main/scala/calc/Typer.scala:
```diff
@@ -0,0 +1,25 @@
+  def binaryOp(t: BinaryOp) = { implicit val pos = t.pos
+    val lhs = inferType(t.lhs)
+    val rhs = inferType(t.rhs)
+    if (lhs.getType() == rhs.getType()) {
```
That's not strict enough. A binary op demands that both sides be doubles. So if both sides are functions, it's still a type error.

evansb replied:
Ah yes, I missed that thanks

sjrd commented on src/main/scala/calc/Typer.scala:
```diff
@@ -0,0 +1,25 @@
+    val rhs = inferType(t.rhs)
+    if (lhs.getType() == rhs.getType()) {
+      new BinaryOpT(t.op, lhs, rhs) {
+        def getType() = lhs.getType()
```
Using a `def` here is dangerous from a complexity point of view. If your lhs is itself a `BinaryOpT`, and recursively so, calling a single `getType` will take liner time, and then `binaryOp()` itself will blow up exponentially.

evansb replied:
Ah yes, I understand.

evansb committed:
Implement Typer phase for Let binding

evansb committed:
Use values instead of methods for typed AST contract

evansb commented:
Allright, I think I resolved all your comments above. Thanks for the quick response :)

Going to implement the rest now.

sjrd commented on src/main/scala/calc/Typer.scala:
```diff
@@ -0,0 +1,33 @@
+
+object Typer {
+
+  type TypeEnv = Map[Ident, Type]
```
Relying on equality of trees is risky. Typically, an environment is a map form `String` to `Type`, where the string is the `name` of an `Ident`, not an `Ident` itself.

evansb replied:
Yes, you are right

evansb committed:
Implement typing rule of Ident which is missing

evansb committed:
Change key of type environment to String

evansb committed:
Implement let binding + positive test

evansb committed:
Fix bug in test case and add several negative test cases for let

evansb committed:
Implement phase typer for IfElse

evansb committed:
Implement If expression

evansb committed:
Implement typer for lambda expression

evansb committed:
Implement phase typer for function call

evansb committed:
Add more negative test now that closure type is available

evansb committed:
Implemented closure and call compiler, but somehow it doesn't work

evansb commented:
Hmm, I think I'm stuck.
I got this error when trying to execute my test case for Closure application which I couldn't decipher

Are we supposed to use `JSFunctionApply` instruction for function call?

```
java.lang.RuntimeException: There were 1 IR checking errors.

    at scala.sys.package$.error(package.scala:27)
    at org.scalajs.core.tools.linker.frontend.BaseLinker$$anonfun$linkInternal$2.apply$mcV$sp(BaseLinker.scala:133)
    at org.scalajs.core.tools.linker.frontend.BaseLinker$$anonfun$linkInternal$2.apply(BaseLinker.scala:130)
    at org.scalajs.core.tools.linker.frontend.BaseLinker$$anonfun$linkInternal$2.apply(BaseLinker.scala:130)
    at org.scalajs.core.tools.logging.Logger$class.time(Logger.scala:28)
    at org.scalajs.core.tools.logging.NullLogger$.time(NullLogger.scala:3)
    at org.scalajs.core.tools.linker.frontend.BaseLinker.linkInternal(BaseLinker.scala:129)
    at org.scalajs.core.tools.linker.frontend.BaseLinker.linkInternal(BaseLinker.scala:85)
    at org.scalajs.core.tools.linker.frontend.LinkerFrontend$$anonfun$4.apply(LinkerFrontend.scala:54)
    at org.scalajs.core.tools.linker.frontend.LinkerFrontend$$anonfun$4.apply(LinkerFrontend.scala:54)
    at org.scalajs.core.tools.logging.Logger$class.time(Logger.scala:28)
    at org.scalajs.core.tools.logging.NullLogger$.time(NullLogger.scala:3)
    at org.scalajs.core.tools.linker.frontend.LinkerFrontend.link(LinkerFrontend.scala:53)
    at org.scalajs.core.tools.linker.Linker$$anonfun$link$1.apply$mcV$sp(Linker.scala:48)
    at org.scalajs.core.tools.linker.Linker$$anonfun$link$1.apply(Linker.scala:47)
    at org.scalajs.core.tools.linker.Linker$$anonfun$link$1.apply(Linker.scala:47)
    at org.scalajs.core.tools.linker.Linker.guard(Linker.scala:65)
    at org.scalajs.core.tools.linker.Linker.link(Linker.scala:47)
    at calc.Linker$.link(Linker.scala:52)
    at calc.TestHelpers$.assertRun(TestHelpers.scala:23)
    at calc.RunSimpleValue.closurePlusApply(RunTest.scala:48)

Process finished with exit code 25
```

sjrd commented:
> java.lang.RuntimeException: There were 1 IR checking errors.

It would be more useful to show the IR checking error which is printed just above ;)

evansb commented:
Strangely it doesn't show anything though, `java.lang.RuntimeException: There were 1 IR checking errors.` is the first line of the log. Do I need to turn on some debugging flag?

I tried both in `sbt` and and IDE.

evansb commented:
Oh, my bad, just realised that the logger in the linker for Testing is using `NullLogger`.

I'll investigate this again

evansb committed:
Enable logging at debug level for testing

evansb committed:
Rename class

sjrd commented:
Also, if you're stuck, make sure to hang out in https://gitter.im/scala-js/scala-js/core
There was some discussion about the assignment, and you can ask guidance there too.

evansb commented:
Noted, thanks :)

evansb committed:
Finish Closure & Apply without capture

evansb committed:
Fix some inconsistent naming in test variable name

evansb committed:
Implement importing foreign functions

evansb commented:
Finished all compulsory tasks except capturing, would you like to give feedback.

I'm a little behind on black box testing though, will add more after I finished the last task.

evansb committed:
Do not hard-code the mangled function name

sjrd commented on src/main/scala/calc/Typer.scala:
```diff
@@ -0,0 +1,97 @@
+  def closure(t: Closure)(implicit env: TypeEnv) = { implicit val pos = t.pos
+    val params = t.params map { p => new IdentT(p.name) { val tpe = TDouble } }
+    val body = inferType(t.body)(env ++ (t.params map { _.name -> TDouble }))
+    new ClosureT(params, body) { val tpe = TFun(params.length) }
```
You should check that `body.tpe` is a `TDouble`, since a closure must always return a number.

sjrd commented on src/main/scala/calc/Compiler.scala:
```diff
@@ -62,21 +61,101 @@ object Compiler {
+    val paramUnbox = t.params map { ident =>
+      val varRef = irt.VarRef(irt.Ident(ident.name))(irtpe.AnyType)
+      val varUnbox = irt.Unbox(varRef, 'D')
+      irt.VarDef(irt.Ident(ident.name), irtpe.DoubleType, false, varUnbox)
```
Uh oh. You're using the same `ident.name` for the `ParamDef` and the local `VarDef`. That's bad, as they live in the same scope. I'm surprised it did not blow up already. You should use a different (generated) name for the `ParamDef` and its `VarRef`.

evansb replied:
Whoops yes, my bad habit of copy pasting >.<

sjrd commented:
That's all I have at this point.

evansb committed:
Implement free variables finder

evansb committed:
Mangle closure parameter name

evansb committed:
Implement capturing in closure

evansb committed:
Uncomment accidentally commented test

evansb committed:
Need to unbox foreign call result

evansb commented:
I think I am done implementation-wise, gonna add more black box and negative tests to find potential bugs.

evansb commented:
Should I try to do extra feature afterwards? 
I think adding one type will breaks lots of original specification and add the need to implement Hindley-Milner which is not trivial.

There are several things that I have in mind that doesn't add type but still let me use more IRs
- Mutable variables
- While loop
- Experiment with optimiser hints?

evansb committed:
Fix parser bug

sjrd commented:
Good! Looks pretty complete now :-) Good job!

There's no need for you to add extra features. But if you want to explore further, here are some ideas:
- Support eta-expansion of math methods, e.g., make this work: `let f = sin in f(1)`, or even `(if (1) sin else cos)(1)`
- Support self-recursion, e.g., support `let fib = fun(n) = { if (n) fib(n-1) * n else 1 } in fib(5)`
- Support booleans and comparison operators between numbers

evansb committed:
Fix typo in test case

evansb committed:
Add more tests

evansb committed:
Implement Eta Expansion on Foreign Functions

evansb commented:
> Support eta-expansion of math methods, e.g., make this work: let f = sin in f(1), or even (if (1) sin else cos)(1)

I implemented this, seems to be working :-)
I added a third phase of the compilation which is to prepend the IR with eta expansions of foreign functions.

And thanks for your second example I found a bug in my foreign function type definition, I did embed foreign function data in the type but forgot to relax the type equality requirement :D

evansb committed:
Foreign Function arity must still be the same

evansb committed:
Initial implementation of Recursive Function

evansb commented:
https://github.com/sjrd/scalajsir-calc/pull/5/commits/7252782bee36772ae2b4be449851a198f94d3491

This changeset and beyond implements recursive function by following your instruction on Gitter.
I managed to run your factorial function example.

It turns out the typechecker become much more complicated because of this since the type of recursive function is not known initially. I apply a type inference algorithm (I think it's called W algorithm). Also need to handle cases like `let f = fun (x) { f(3) + f(4, 5) }`

Obviously since this is extra feature it might not as robust as compulsory features (though it doesn't seem to break existing test cases so far). Let me know if you want me to split it into separate PR :)

evansb committed:
Allow recursive function to be shadowed by let expression inside

evansb committed:
Ensure parameters are unique in phase typer

sjrd commented on src/main/scala/calc/Foreign.scala:
```diff
@@ -0,0 +1,38 @@
+          val params = (1 to t.arity).map({ n => Ident("a__" + n) }).toList
+          val closure = Closure(params, Call(Ident(name), params))
+          val (typedClosure, _) = Typer.closure(closure)
+          val compiledClosure = Compiler.closure(typedClosure)
```
Ah ah! That's funny. Desugaring in your source language then compiling that with your existing compiler.

sjrd commented on src/main/scala/calc/Tree.scala:
```diff
@@ -1,6 +1,30 @@
+  override def equals(o: Any) = o match {
+    case that: TStaticForeignFun => this.arity == that.arity
+    case _ => false
+  }
```
That `equals`method is suspicious. It might indicate that there is something unelegant about encoding `clsName` and `method` in the _type_ of foreign functions.

Anyway, if you redefine `equals`, you must also redefine `hashCode`.

evansb replied:
Yeah, ideally foreign functions information should be stored elsewhere, not encoded in the type.

This was implemented before my eta reduce so that I don't need to include environment as additional parameter in my compiler code.

sjrd commented on src/main/scala/calc/Compiler.scala:
```diff
@@ -62,21 +61,212 @@ object Compiler {
+
+    // Phase 3: Prepend Eta expansion of foreign method to IR
+    if (etaExpansion) {
+      Foreign.prependEtaExpansion(ir, foreignEnv)
```
Ideally, you would only prepend the eta-expansion of functions that are actually used in the `typedAst` :-)
Detecting this is similar to identify free variables.

evansb replied:
Yep, this optimisation can be added trivially, just get the free variables of the whole program and only prepend them. At this stage free variables implies it's a foreign function.

sjrd commented:
Note that, in fact, there exists a simpler implementation of eta-expansion, which does not need to figure out which ones are used, nor prepend something at the beginning of the program. It is simply to generate a closure for the compilation of an `Ident` if that `Ident` happens to refer to a foreign function. Note that you would not even get there if the foreign function is used directly in a `Call`, since you do not recursively translate the `Ident` in that case.

evansb commented:
Hmm yeah I think that's simpler. Again, similar with the reasoning behind embedding method and class name in type, this will be possible if we include environment as additional parameter in the compiler

evansb commented:
By the way, can I make revisions? I think your suggestion is trivial to implement

sjrd commented:
> By the way, can I make revisions? I think your suggestion is trivial to implement

Sure, sure.

sjrd commented on src/main/scala/calc/Compiler.scala:
```diff
@@ -62,21 +61,212 @@ object Compiler {
+        val emptyObject = irt.JSObjectConstr(List())
+        val bindingName = t.name.name
+        val objectName = irt.Ident(bindingName + "__ref")
+        val objectDef = irt.VarDef(objectName, irtpe.AnyType, false, emptyObject)
```
I see you're using a simple JS object as "ref box". This is not the best. In the Scala standard library, there are classes specially dedicated for these boxes: `scala.runtime.XYZRef`. In particular, you will want a `scala.runtime.ObjectRef`, whose field has type `any`. You can create a box which is initially empty with the "static" method `ObjectRef$.zero()`.

The advantage of using the `Ref` classes compare to `JSObjectConstr` is that the optimizer can understand it, since it is a Scala class. In some cases, that means the optimizer can completely eliminate the box after some other optimizations are performed.

sjrd replied:
You can find the definitions of the `Ref` classes here: https://github.com/scala-js/scala-js/blob/master/library-aux/src/main/scala/scala/runtime/RefTypes.scala

sjrd commented on src/main/scala/calc/Compiler.scala:
```diff
@@ -62,21 +61,212 @@ object Compiler {
+    val receiver = irt.LoadModule(irtpe.ClassType(encodeClassName(t.clsName + "$")))
+    val method = irt.Ident(t.methodName)
+    val args = t.args map expr
+    irt.Unbox(irt.Apply(receiver, method, args)(t.tpe.irtype), 'D')
```
That was probably there before, but I notice now that the `Unbox` is not needed for an `Apply` node. That is because "foreign" calls, as you call them, are actually the _normal_ calls of Scala.js, which are statically typed. `JSFunctionApply` nodes are the "foreign" calls of the Scala.js IR, as they call JavaScript functions, which are dynamically typed.

sjrd commented:
> It turns out the typechecker become much more complicated because of this since the type of recursive function is not known initially. I apply a type inference algorithm (I think it's called W algorithm).

That sounds much overkill. We know that functions always have double parameters and return doubles. You do not have to typecheck the body of a function to already know that it will return a `Double`, so you precisely know the type of the function before you typecheck its body. There should be no need to complicate the type checker.

evansb committed:
Simplify foreign function interface

evansb committed:
Fix several bugs

evansb committed:
Use ObjectRef instead of JS object

evansb committed:
Revert back to simpler typer with function type assumptions

evansb commented:
All right, I think I'm caught up with your suggestions :). Thanks for all your feedback and suggestions so far, it really helped me understanding the IR

evansb committed:
Remove obsolete comment

sjrd commented:
I am on holiday until next Sunday, and won't be able to have a look at it until then.
